slsReceiverSoftware/src/FrameSynchronizerApp.cpp

```diff
@@ -0,0 +1,575 @@
+#include <unistd.h>
+
+#include <set>
+#include <vector>
+#include <zmq.h>
+
+std::vector<std::thread> threads;
+std::vector<sem_t *> semaphores;
+sls::TLogLevel printHeadersLevel = sls::logDEBUG;
+
+/** Define Colors to print data call back in different colors for different
+ * recievers */
+#define PRINT_IN_COLOR(c, f, ...)                                              \
+    printf("\033[%dm" f RESET, 30 + c + 1, ##__VA_ARGS__)
+
+/** Structure handling different threads */
+struct Status {
+    bool starting = true;
+    bool terminate = false;
+    unsigned long num_receivers;
+    sem_t available;
+    std::mutex mtx;
+    std::vector<zmq_msg_t *> headers;
+    std::map<unsigned int,
+             std::map<long unsigned int, std::vector<zmq_msg_t *>>>
+        frames;
+    std::vector<zmq_msg_t *> ends;
+
+    Status(bool start, bool term, unsigned long num_recv)
+        : starting(start), terminate(term), num_receivers(num_recv) {
+        sem_init(&available, 0, 0);
+    }
+};
+
+/**
+ * Control+C Interrupt Handler
+ * to let all the processes know to exit properly
+ */
+void sigInterruptHandler(int p) {
+    for (size_t i = 0; i != semaphores.size(); ++i) {
+        sem_post(semaphores[i]);
+    }
+}
+
+/**
+ * prints usage of this example program
+ */
+std::string getHelpMessage() {
+    std::ostringstream os;
+    os << "\nUsage:\n"
+          "./slsFrameSynchronizer [start tcp port] [num recevers] [print "
+          "callback headers (optional)]\n"
+       << "\t - tcp port has to be non-zero and 16 bit\n"
+       << "\t - print callback headers option is 0 (disabled) by default\n";
+    return os.str();
+}
+
+void zmq_free(void *data, void *hint) { free(data); }
+
+void print_frames(
+    const std::map<unsigned int,
+                   std::map<long unsigned int, std::vector<zmq_msg_t *>>>
+        &frames) {
+    LOG(sls::logDEBUG) << "Printing frames";
+    for (const auto &outer_pair : frames) {
+        unsigned int udpPort = outer_pair.first;
+        const auto &trigger_map = outer_pair.second;
+        LOG(sls::logDEBUG) << "UDP port: " << udpPort;
+        for (const auto &inner_pair : trigger_map) {
+            long unsigned int acqIndex = inner_pair.first;
+            const auto &msg_vector = inner_pair.second;
+            LOG(sls::logDEBUG) << "  acq index: " << acqIndex << '['
+                               << msg_vector.size() << ']';
+        }
+    }
+}
+
+std::set<long unsigned int>
+find_keys(const std::map<unsigned int,
+                         std::map<long unsigned int, std::vector<zmq_msg_t *>>>
+              &maps) {
+    std::set<long unsigned int>
+        all_keys; // Set to collect all unique keys across all maps
+    std::set<long unsigned int> valid_keys; // Set to store final valid keys
+
+    // If no maps are provided, return empty set
+    if (maps.empty()) {
+        return valid_keys;
+    }
+
+    // Collect all unique keys from all maps
+    for (const auto &[port, trigger_map] : maps) {
```

Comment:
Structured bindings requires c++17. 

This would break compilation on RH7 (gcc4.8). Which is ok but I guess we need to inform the group. This definitely makes in a major version.